### PR TITLE
Sample App - Avoid showing consent screen twice for same pending/push challenge

### PR DESCRIPTION
**Pull request by moisesolmedo-okta**
Checking by transactionID before presenting the user consent screen

### CODE REVIEW

File: Examples/PushSampleApp/SampleApp/RootCoordinator.swift [hunk 1]
```diff
@@ -95,14 +95,27 @@ class RootCoordinator {
             nav.dismiss(animated: true)
         }
         userConsentVC.viewModel = viewModel
-        if nav.presentedViewController != nil {
-            nav.dismiss(animated: true) {
+        if !isPresentingSameChallenge(nav: nav, newStep: remediationStep) {
```
Comment (IldarAbdullin-okta) [inline Examples/PushSampleApp/SampleApp/RootCoordinator.swift:98]: Why even ask sdk to resolve challenge if we can drop early after `parseChallenge` step?

**Review by IldarAbdullin-okta — CHANGES_REQUESTED** [review]
It is not recommended to start `resolve` flow and abandon it. It is better to drop challenge early after `parsing` step

File: Examples/PushSampleApp/SampleApp/RootCoordinator.swift (hunk 1 above)
Comment (moisesolmedo-okta) [reply Examples/PushSampleApp/SampleApp/RootCoordinator.swift]: good point. Changed it to drop after parsing the challenge.

File: Examples/PushSampleApp/SampleApp/PushNotificationService.swift [hunk 2]
```diff
@@ -133,6 +134,8 @@ class PushNotificationService: NSObject, UNUserNotificationCenterDelegate {
     }
 
     private func resolvePushChallenge(_ pushChallenge: PushChallengeProtocol) {
+        guard !(delegate?.isPresentingSameChallenge(newStep: pushChallenge) ?? false) else { return }
```
Comment (IldarAbdullin-okta) [inline Examples/PushSampleApp/SampleApp/PushNotificationService.swift:137]: Still looks complicated. Why just not create queue/array that keeps unhandled/in-progress challenges? In that case you can use array to find whether challenge is unique or duplicate

Comment (IldarAbdullin-okta) [reply Examples/PushSampleApp/SampleApp/PushNotificationService.swift:137]: You can kick-out challenges in `onCompletion` closure below

File: Examples/PushSampleApp/SampleApp/PushNotificationService.swift [hunk 3]
```diff
@@ -133,20 +134,33 @@ class PushNotificationService: NSObject, UNUserNotificationCenterDelegate {
...
     private func resolvePushChallenge(_ pushChallenge: PushChallengeProtocol) {
+        guard shouldHandleChallenge(newChallenge: pushChallenge) else { return }
+        unhandledChallenges.insert(pushChallenge.transactionId)
+
         pushChallenge.resolve { step in
             self.remediationEventsHandler.handle(step: step)
-        } onCompletion: { error in
+        } onCompletion: { [weak self] error in
+            guard let self = self else { return }
+            self.unhandledChallenges.remove(pushChallenge.transactionId)
```
Comment (IldarAbdullin-okta) [inline Examples/PushSampleApp/SampleApp/PushNotificationService.swift:144]: Just want to make sure that we inserting and removing on UI thread, correct?

File: Examples/PushSampleApp/SampleApp/PushNotificationService.swift [hunk 4]
```diff
@@ -133,20 +134,33 @@ class PushNotificationService: NSObject, UNUserNotificationCenterDelegate {
...
             DispatchQueue.main.async {
                 if let error = error {
                     self.logger?.error(eventName: LoggerEvent.account.rawValue, message: "Cannot resolve push challenge: \(error.localizedDescription)")
                 } else {
                     self.remediationEventsHandler.onChallengeResolved(pushChallenge.userResponse)
-                    self.logger?.error(eventName: LoggerEvent.account.rawValue, message: "Success resolving push challenge")
+                    self.logger?.info(eventName: LoggerEvent.account.rawValue, message: "Success resolving push challenge")
```
Comment (IldarAbdullin-okta) [inline Examples/PushSampleApp/SampleApp/PushNotificationService.swift:150]: 👍

**Review by IldarAbdullin-okta — APPROVED** [review]
Nice! 👍

File: Examples/PushSampleApp/SampleApp/PushNotificationService.swift (hunk 3 above)
Comment (moisesolmedo-okta) [reply Examples/PushSampleApp/SampleApp/PushNotificationService.swift]: Turns out the completion was being dispatched to a different NSURLSession queue. Moved the removal to UI thread.

**Pull request merged**